src/EFCore.PG/Query/Internal/NpgsqlSqlTranslatingExpressionVisitor.cs

```diff
@@ -246,6 +253,52 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCall)
                 }
             }
 
+            {
+                // Pattern match for: array.Any(e => e == x) (and other equality patterns)
+                // Transform this to Contains.
+                if (method.IsClosedFormOf(EnumerableAnyWithPredicate) &&
+                    arguments[1] is LambdaExpression wherePredicate &&
+                    TryMatchEquality(wherePredicate.Body, out var left, out var right) &&
+                    (left == wherePredicate.Parameters[0] || right == wherePredicate.Parameters[0]))
+                {
+                    var item = left == wherePredicate.Parameters[0]
+                        ? right
+                        : right == wherePredicate.Parameters[0]
+                            ? left
+                            : null;
+
+                    return item is null
+                        ? null
+                        : Visit(Expression.Call(Contains.MakeGenericMethod(method.GetGenericArguments()[0]), array, item));
+                }
+
+                static bool TryMatchEquality(Expression expression, out Expression left, out Expression right)
+                {
+                    if (expression is BinaryExpression binary)
+                    {
+                        (left, right) = (binary.Left, binary.Right);
+                        return true;
+                    }
+                    if (expression is MethodCallExpression methodCall)
+                    {
+                        if (methodCall.Method == ObjectEquals)
+                        {
+                            (left, right) = (methodCall.Arguments[0], methodCall.Arguments[1]);
+                            return true;
+                        }
+
+                        if (methodCall.Method.Name == nameof(object.Equals) && methodCall.Arguments.Count == 1)
+                        {
+                            (left, right) = (methodCall.Object, methodCall.Arguments[0]);
```

Comment:
Looks like this method produces sorted results already, so why not to determine ordering for the static equality method and for binary expressions?

Reply:
I don't understand... The only job of TryMatchEquality is to provide the two side of the equality - there's no sorting/ordering involved...

Reply:
I meant this:

```csharp
                    var item = left == wherePredicate.Parameters[0]
                        ? right
                        : right == wherePredicate.Parameters[0]
                            ? left
                            : null;
```

It can be made simpler probably if it will combined with that method.

Reply:
TryMatchEquality is just about identifying equality and returning its parameters - it could be extracted and used elsewhere too. But if you think code can be really simplified here, do you want to submit a full code proposal?

Reply:
Not today, you caught me while I was heading to bed (:

Reply:
No problem for it to come after merge too :)